library/core/src/iter/adapters/intersperse.rs

```diff
@@ -0,0 +1,72 @@
+use crate::iter::Fuse;
+
+/// An iterator adapter that places a separator between all elements.
+#[unstable(feature = "iter_intersperse", reason = "recently added", issue = "none")]
+#[derive(Debug, Clone)]
+pub struct Intersperse<I: Iterator>
+where
+    I::Item: Clone,
+{
+    element: I::Item,
+    iter: Fuse<I>,
+    peek: Option<I::Item>,
+}
+
+impl<I: Iterator> Intersperse<I>
+where
+    I::Item: Clone,
+{
+    pub(in crate::iter) fn new(iter: I, separator: I::Item) -> Self {
+        let mut iter = iter.fuse();
+        Self { peek: iter.next(), iter, element: separator }
```

Comment:
Is that a good idea? Then merely constructing the iterator will advance the wrapped one. `Peekable` which is similar doesn't do this.
It violates [laziness](https://doc.rust-lang.org/std/iter/index.html#laziness) property of iterator adapters.